README.md
# SeinfeldTrivia

Project started sometime in 2016, and fixed up a little bit on 03/18/19.

A Seinfeld trivia application. Add your own questions in the data/data.txt file!


docs/changelog.txt
Seinfeld Trivia Changelog
=========================

LEGEND:
- 'i' stands for 'in-development', the stage before Alpha, not very functional and lots o' bugs
- 'a' stands for 'alpha', the early but mostly functional stage, somewhat buggy
- 'b' stands for 'beta', the more polished but not finished yet stage, small amount of bugs
- 'r' stands for 'release', the final stage where the software is not in active development

VERSION: 0.1i - 10/8/16
-----------------------
- In-development

- Bugginess Level: High 

- Created program "shell"
- Added seinfeld logo with "TRIVIA!" text
- Added "Start Trivia" button that took up half the dang screen

VERSION: 0.2i - 10/9/16
-----------------------
- Bugginess Level: Medium

- Switched HomeGui JPanel layout from GridLayout
to BoxLayout
- Fixed the "Start Trivia" button
- Changed "look and feel" of GUI to system's/OS's look and feel
(Ubuntu will have Ubuntu look, Windows will have Windows look,
macOS will have macOS look)
- Added "Options" button
- Added "Quit" button
- Added basic button functionality
- Changed some initialization structure
- Added Options JPanel "screen" with an experimental picture

VERSION: 0.4i - 10/29/16
------------------------
- Bugginess Level: High

- Implemented data-gathering system
- Implemented QuestionGui basics
- Implemented Correct Answer system for QuestionGui
- Implemented one-line, on their own line, comments for data.txt file 
- Implemented proper button functionality (when correct answer or wrong
answer buttons are clicked, they show their respective colors and all the
other buttons are disabled)
- Implemented a (by default) 5 second timer that gives the user time
to see if they got the question right or wrong before moving on to the
next question, program moves on to next question automatically
- Made data gathering and the triviaData ArrayList transfer from Data
to QuestionGui more efficient (Data no longer repeatedly goes through
the data.txt file after doing it once)
- Fixed several issues related to previous entry

VERSION: 0.5a - 11/12/16
------------------------
- ALPHA!

- Bugginess Level: Medium

- Cleaned up the infamous QuestionGui actionPerformed method
- Implemented two buggy timers (the answer timer and the move on delay timer)
- Fixed a timer bug
- All timers work! (mostly)
- The timer will stop and not display 'out of time!' or anything else when
user pressed the correct answer right when time is 0
- Implemented a thread that continuously checks if user has pressed the 'start trivia'
button (which is synonymous with arriving at the QuestionGui JPanel/'screen'. If so, the
answer timer is started. This prevents the timer running and running out of time before
the user ever enters the QuestionGui JPanel

VERSION: 0.7a - 11/22/16
------------------------
- Bugginess Level: Low

- Significant improvements were made in the previous version,
so the program was promoted directly to 0.7, skipping 0.6
- Implemented question limit, default number of questions is 10
- Fixed one of the answer timer bugs
- Implemented the ScoreboardGui

VERSION: 0.8b - 11/26/16
------------------------
- BETA!

- Bugginess Level: Low

- All the main features designated from the start have been
implemented. Now, code clean-up & organization and documentation/commenting
will be the focus of the Beta phase. Major Aesthetic work will probably not be
implemented, ever.

src/elaguy/personal/seinfeldtrivia/main/Data.java
package elaguy.personal.seinfeldtrivia.main;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.StandardCopyOption;
import java.util.ArrayList;

import javax.swing.JOptionPane;

public class Data {
	
	private SeinfeldTrivia seinfeldTrivia;
	private ArrayList<String>triviaData;
	private int questionCount;
	private ArrayList<String>correctAnswers;
	
	public Data(SeinfeldTrivia seinfeldTrivia) {
		reset(seinfeldTrivia);
	}
		
	public void gatherData() {
		String home = System.getProperty("user.home");
		
		String dataDirPath = home + "/Documents/SeinfeldTrivia/";
		File dataDir = new File(dataDirPath);
		
		String dataFileName = "data.txt";
		
		if(!dataDir.exists())
			dataDir.mkdir();
		
		File localDataDir = new File("data");
		File[] dataFiles = localDataDir.listFiles();
		
		for(File file : dataFiles) {
			if(!(new File(dataDirPath + file.getName()).exists())) {
				try {
					Files.copy(file.toPath(), (new File(dataDirPath + file.getName())).toPath());
				} catch (IOException e) {
					JOptionPane.showMessageDialog(seinfeldTrivia.getMainFrame(), 
							"An Input/Output error occurred!\n"
							+ "The program failed to copy critical data files.",
							"IO Error", JOptionPane.ERROR_MESSAGE);
					seinfeldTrivia.getMainFrame().dispose();
					System.exit(0);
				}
			}
		}
		
		BufferedReader br = null;
		try {
			br = new BufferedReader(new FileReader(dataDirPath + dataFileName));
			
			String line = "";
			int index = -1;
			while((line = br.readLine()) != null) {
				line = line.trim();
				if(!line.equals("")) { // ignore whitespaces
					if((line.charAt(0) != '/') && (line.charAt(1) != '/')) { // ignore one-line comments ON THEIR OWN LINE
						for(int i = 0; i < line.length(); i++) {
							if(line.charAt(i) == '+') {
								line = line.substring(0, i) + line.substring(i+1);
								correctAnswers.add(line);
						}
					}
					triviaData.add(line);
					
					if(index % 6 == 0)
						questionCount++;
					}
				index++;
				}
			}
		} 
		
		catch (IOException e) {
			JOptionPane.showMessageDialog(seinfeldTrivia.getMainFrame(), 
					"An Input/Output error occurred!\n"
					+ "This usually means the program could not"
					+ " find a question data file.",
					"IO Error", JOptionPane.ERROR_MESSAGE);
			seinfeldTrivia.getMainFrame().dispose();
			System.exit(0);
		}
		
		finally {
			
			try {
				br.close();
			} catch (IOException e) {
				JOptionPane.showMessageDialog(seinfeldTrivia.getMainFrame(), 
						"An Input/Output error occurred!\n"
						+ " This usually means the program could not"
						+ " find a question data file.",
						"IO Error", JOptionPane.ERROR_MESSAGE);
				seinfeldTrivia.getMainFrame().dispose();
				System.exit(0);
			}
		}
	}
	
	// resets the whole class, equivalent to inside of constructor
	private void reset(SeinfeldTrivia seinfeldTrivia) {
		this.seinfeldTrivia = seinfeldTrivia;
		
		triviaData = new ArrayList<String>();
		questionCount = 0;
		correctAnswers = new ArrayList<String>();
	}
	
	public ArrayList<String> getTriviaData() {
		return triviaData;
	}
	
	public int getQuestionCount() {
		return questionCount;
	}

	public void setQuestionGui(QuestionGui questionGui) {
		
	}
	
	public ArrayList<String> getCorrectAnswers() {
		return correctAnswers;
	}
	
}


src/elaguy/personal/seinfeldtrivia/main/QuestionGui.java
package elaguy.personal.seinfeldtrivia.main;

import java.awt.BorderLayout;
import java.awt.Color;
import java.awt.Component;
import java.awt.Dimension;
import java.awt.Font;
import java.awt.Image;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.util.ArrayList;
import java.util.Random;
import javax.swing.Timer;
import javax.swing.Box;
import javax.swing.BoxLayout;
import javax.swing.ImageIcon;
import javax.swing.JButton;
import javax.swing.JLabel;
import javax.swing.JOptionPane;
import javax.swing.JPanel;
import javax.swing.SwingConstants;

public class QuestionGui extends JPanel implements ActionListener, Runnable {

	private SeinfeldTrivia seinfeldTrivia;
	private Data data;
	private Graphics graphics;
	private HomeGui homeGui;
	private ScoreboardGui scoreboardGui;
	
	private ArrayList<String>triviaData;
	private int questionCount, questionIndex, imgIndex;
	private int boundaryIndex1, boundaryIndex2;
	private ArrayList<String> correctAnswers;
	private String rightAnswer;
	
	private JButton answer1, answer2, answer3, answer4;
	private JLabel timeLeftLabel, questionNumLabel;
	
	private boolean hasAnswered, answerTimerIsStopped;
	
	private int moveOnDelay; // in seconds
	private int timeLeft; // in seconds
	
	private int currentQuestionNum;
	private int numOfQuestions;
	
	private Timer moveOnTimer, answerTimer;
	
	private int numCorrect;
	private int numIncorrect;
	
	
	public QuestionGui(SeinfeldTrivia seinfeldTrivia, Data data, Graphics graphics, 
			HomeGui homeGui, ScoreboardGui scoreboardGui) {
		reset(seinfeldTrivia, data, graphics, homeGui, scoreboardGui);
	}
	
	private void initContent() {
		//System.out.println(questionIndex);
		
		String home = System.getProperty("user.home");
		String dataDirPath = home + "/Documents/SeinfeldTrivia/";
		int answerIndex = new Random().nextInt(boundaryIndex2 - boundaryIndex1) + boundaryIndex1;
		
		JPanel questionPanel = new JPanel(new BorderLayout());
		questionPanel.setMaximumSize(new Dimension(seinfeldTrivia.getWIDTH(), 50));
		
		JLabel question = new JLabel(triviaData.get(questionIndex));
		question.setFont(new Font("Georgia", Font.BOLD, 18));
		question.setVerticalAlignment(SwingConstants.TOP);
		question.setHorizontalAlignment(SwingConstants.CENTER);
		questionPanel.add(question);
		this.add(questionPanel);
		
		timeLeftLabel = new JLabel("Time Left: " + timeLeft);
		timeLeftLabel.setFont(new Font("Georgia", Font.BOLD, 12));
		timeLeftLabel.setAlignmentX(Component.CENTER_ALIGNMENT);
		this.add(Box.createRigidArea(new Dimension(0, 5)));
		this.add(timeLeftLabel);
		
		questionNumLabel = new JLabel("Question " + currentQuestionNum + " of " + numOfQuestions);
		questionNumLabel.setFont(new Font("Georgia", Font.BOLD, 12));
		questionNumLabel.setAlignmentX(Component.CENTER_ALIGNMENT);
		this.add(Box.createRigidArea(new Dimension(0, 10)));
		this.add(questionNumLabel);
		
		JLabel image = new JLabel();
		ImageIcon imageIcon = new ImageIcon(dataDirPath + triviaData.get(imgIndex));
		Image newImg = graphics.getResizedImage(imageIcon.getImage(), 200, 200);
		imageIcon = new ImageIcon(newImg);
		image.setIcon(imageIcon);
		image.setAlignmentX(Component.CENTER_ALIGNMENT);
		this.add(Box.createRigidArea(new Dimension(0, 25)));
		this.add(image);
		
		answer1 = new JButton(triviaData.get(answerIndex));
		answer1.setAlignmentX(Component.CENTER_ALIGNMENT);
		answer1.setMinimumSize(new Dimension(75, 10));
		answer1.setMaximumSize(new Dimension(150, 40));
		answer1.setForeground(Color.RED);
		answer1.setOpaque(true);
		answer1.addActionListener(this);
		answer1.setActionCommand("answer1");
		this.add(Box.createRigidArea(new Dimension(0, 30)));
		this.add(answer1);
		triviaData.remove(answerIndex);
		
		answerIndex = new Random().nextInt((boundaryIndex2-=1) - boundaryIndex1) + (boundaryIndex1);
		answer2 = new JButton(triviaData.get(answerIndex));
		answer2.setAlignmentX(Component.CENTER_ALIGNMENT);
		answer2.setMinimumSize(new Dimension(75, 10));
		answer2.setMaximumSize(new Dimension(150, 40));
		answer2.setForeground(Color.RED);
		answer2.setOpaque(true);
		answer2.addActionListener(this);
		answer2.setActionCommand("answer2");
		this.add(Box.createRigidArea(new Dimension(0, 45)));
		this.add(answer2);
		triviaData.remove(answerIndex);
		
		answerIndex = new Random().nextInt((boundaryIndex2-=1) - boundaryIndex1) + (boundaryIndex1);
		answer3 = new JButton(triviaData.get(answerIndex));
		answer3.setAlignmentX(Component.CENTER_ALIGNMENT);
		answer3.setMinimumSize(new Dimension(75, 10));
		answer3.setMaximumSize(new Dimension(150, 40));
		answer3.setForeground(Color.RED);
		answer3.setOpaque(true);
		answer3.addActionListener(this);
		answer3.setActionCommand("answer3");
		this.add(Box.createRigidArea(new Dimension(0, 45)));
		this.add(answer3);
		triviaData.remove(answerIndex);
		
		answerIndex = new Random().nextInt((boundaryIndex2-=1) - boundaryIndex1) + (boundaryIndex1);
		answer4 = new JButton(triviaData.get(answerIndex));
		answer4.setAlignmentX(Component.CENTER_ALIGNMENT);
		answer4.setMinimumSize(new Dimension(75, 10));
		answer4.setMaximumSize(new Dimension(150, 40));
		answer4.setForeground(Color.RED);
		answer4.setOpaque(true);
		answer4.addActionListener(this);
		answer4.setActionCommand("answer4");
		this.add(Box.createRigidArea(new Dimension(0, 45)));
		this.add(answer4);
		triviaData.remove(answerIndex);
		
		seinfeldTrivia.getMainFrame().pack();
	}
	
	public void actionPerformed(ActionEvent e) {
		if(e.getActionCommand().equals("answer1")) {
			if(answer1.getText().equals(rightAnswer)) {
				if(!hasAnswered) {
					numCorrect++;
					modifyAnswer1AsCorrect();
				}
			}
			
			else {
				if(!hasAnswered) {
					numIncorrect++;
					modifyAnswer1AsWrong();
				}
			}
		}
		
		else if(e.getActionCommand().equals("answer2")) {
			if(answer2.getText().equals(rightAnswer)) {
				if(!hasAnswered) {
					numCorrect++;
					modifyAnswer2AsCorrect();
				}
			}
			
			else {
				if(!hasAnswered) {
					numIncorrect++;
					modifyAnswer2AsWrong();
				}
			}
		}
		
		else if(e.getActionCommand().equals("answer3")) {
			if(answer3.getText().equals(rightAnswer)) {
				if(!hasAnswered) {
					numCorrect++;
					modifyAnswer3AsCorrect();
				}
			}
			
			else {
				if(!hasAnswered) {
					numIncorrect++;
					modifyAnswer3AsWrong();
				}
			}
		}
		
		else if(e.getActionCommand().equals("answer4")) {
			if(answer4.getText().equals(rightAnswer)) {
				if(!hasAnswered) {
					numCorrect++;
					modifyAnswer4AsCorrect();
				}
			}
			
			else {
				if(!hasAnswered) {
					numIncorrect++;
					modifyAnswer4AsWrong();
				}
			}
		}
		
	}
	
	private void modifyAnswer1AsCorrect() {
		answer1.setBackground(Color.GREEN);
		answer1.setForeground(Color.GREEN);
		answer1.setFocusable(false);
		
		answer2.setForeground(Color.GRAY);
		answer2.setEnabled(false);
		answer2.setFocusable(false);
		
		answer3.setForeground(Color.GRAY);
		answer3.setEnabled(false);
		answer3.setFocusable(false);
		
		answer4.setForeground(Color.GRAY);
		answer4.setEnabled(false);
		answer4.setFocusable(false);
		
		hasAnswered = true;
		
		waitThenMoveOn(moveOnDelay);
	}
	
	private void modifyAnswer1AsWrong() {
		answer1.setBackground(Color.RED);
		answer1.setForeground(Color.RED);
		answer1.setFocusable(false);
		
		answer2.setForeground(Color.GRAY);
		answer2.setEnabled(false);
		answer2.setFocusable(false);
		
		answer3.setForeground(Color.GRAY);
		answer3.setEnabled(false);
		answer3.setFocusable(false);
		
		answer4.setForeground(Color.GRAY);
		answer4.setEnabled(false);
		answer4.setFocusable(false);
		
		hasAnswered = true;
		
		waitThenMoveOn(moveOnDelay);
	}
	
	private void modifyAnswer2AsCorrect() {
		answer2.setBackground(Color.GREEN);
		answer2.setForeground(Color.GREEN);
		answer2.setFocusable(false);
		
		answer1.setForeground(Color.GRAY);
		answer1.setEnabled(false);
		answer1.setFocusable(false);
		
		answer3.setForeground(Color.GRAY);
		answer3.setEnabled(false);
		answer3.setFocusable(false);
		
		answer4.setForeground(Color.GRAY);
		answer4.setEnabled(false);
		answer4.setFocusable(false);
		
		hasAnswered = true;
		
		waitThenMoveOn(moveOnDelay);
	}
	
	private void modifyAnswer2AsWrong() {
		answer2.setBackground(Color.RED);
		answer2.setForeground(Color.RED);
		answer2.setFocusable(false);
		
		answer1.setForeground(Color.GRAY);
		answer1.setEnabled(false);
		answer1.setFocusable(false);
		
		answer3.setForeground(Color.GRAY);
		answer3.setEnabled(false);
		answer3.setFocusable(false);
		
		answer4.setForeground(Color.GRAY);
		answer4.setEnabled(false);
		answer4.setFocusable(false);
		
		hasAnswered = true;
		
		waitThenMoveOn(moveOnDelay);
	}
	
	private void modifyAnswer3AsCorrect() {
		answer3.setBackground(Color.GREEN);
		answer3.setForeground(Color.GREEN);
		answer3.setFocusable(false);
		
		answer1.setForeground(Color.GRAY);
		answer1.setEnabled(false);
		answer1.setFocusable(false);
		
		answer2.setForeground(Color.GRAY);
		answer2.setEnabled(false);
		answer2.setFocusable(false);
		
		answer4.setForeground(Color.GRAY);
		answer4.setEnabled(false);
		answer4.setFocusable(false);
		
		hasAnswered = true;
		
		waitThenMoveOn(moveOnDelay);
	}
	
	private void modifyAnswer3AsWrong() {
		answer3.setBackground(Color.RED);
		answer3.setForeground(Color.RED);
		answer3.setFocusable(false);
		
		answer1.setForeground(Color.GRAY);
		answer1.setEnabled(false);
		answer1.setFocusable(false);
		
		answer2.setForeground(Color.GRAY);
		answer2.setEnabled(false);
		answer2.setFocusable(false);
		
		answer4.setForeground(Color.GRAY);
		answer4.setEnabled(false);
		answer4.setFocusable(false);
		
		hasAnswered = true;
		
		waitThenMoveOn(moveOnDelay);
	}
	
	private void modifyAnswer4AsCorrect() {
		answer4.setBackground(Color.GREEN);
		answer4.setForeground(Color.GREEN);
		answer4.setFocusable(false);
		
		answer1.setForeground(Color.GRAY);
		answer1.setEnabled(false);
		answer1.setFocusable(false);
		
		answer2.setForeground(Color.GRAY);
		answer2.setEnabled(false);
		answer2.setFocusable(false);
		
		answer3.setForeground(Color.GRAY);
		answer3.setEnabled(false);
		answer3.setFocusable(false);
		
		hasAnswered = true;
		
		waitThenMoveOn(moveOnDelay);
	}
	
	private void modifyAnswer4AsWrong() {
		answer4.setBackground(Color.RED);
		answer4.setForeground(Color.RED);
		answer4.setFocusable(false);
		
		answer1.setForeground(Color.GRAY);
		answer1.setEnabled(false);
		answer1.setFocusable(false);
		
		answer2.setForeground(Color.GRAY);
		answer2.setEnabled(false);
		answer2.setFocusable(false);
		
		answer3.setForeground(Color.GRAY);
		answer3.setEnabled(false);
		answer3.setFocusable(false);
		
		hasAnswered = true;
		
		waitThenMoveOn(moveOnDelay);
	}
	
	/*  User has answerTime (default is 10) seconds to answer the question.
	 *  If the question is still unanswered at the end of the timer,
	 *  the correct answer is shown and the program moves on to the next
	 *  question
	 */
	private void runAnswerTimer() {
		ActionListener answerTimerListener = new ActionListener() {
			
			@Override
			public void actionPerformed(ActionEvent e) {
				if(timeLeft <= 0 && !hasAnswered) { // if time is up and user has not answered
					answerTimer.stop();
					answerTimerIsStopped = true;
					JOptionPane.showMessageDialog(seinfeldTrivia.getMainFrame(),
							"Time is up!");
					getRightAnswerAndModifyButtons();
				}
				
				else if(timeLeft <= 0 && hasAnswered) { // if time is up but user has answered at the last second
					waitThenMoveOn(moveOnDelay);
				}
				
				else if(!hasAnswered) { // if time is not up and user has not answered
					timeLeft--;
					timeLeftLabel.setText("Time Left: " + timeLeft);
				}
			}
			
		};
		
		answerTimer = new Timer(1000, answerTimerListener);
		answerTimer.start();
	}                                       
	
	// Is used when user clicks an answer button
	private void waitThenMoveOn(int sec) { // in seconds
		correctAnswers.remove(rightAnswer);
		
		if(!answerTimerIsStopped)
			answerTimer.stop();
		
		ActionListener moveOnListener = new ActionListener() {

			@Override
			public void actionPerformed(ActionEvent e) {
				moveOnTimer.stop();
				moveToNextQuestion();
			}
			
		};
		
		moveOnTimer = new Timer(moveOnDelay * 1000, moveOnListener);
		moveOnTimer.start();
	}
	
	// clears and re-populates QuestionGui with next question
	private void moveToNextQuestion() {
		if(currentQuestionNum < numOfQuestions) {
			this.removeAll();
			this.updateUI();
			
			seinfeldTrivia.getMainFrame().repaint();
			seinfeldTrivia.getMainFrame().revalidate();
			
			triviaData.remove(questionIndex); // avoid having same question twice
			triviaData.remove(questionIndex);
			
			questionCount -= 1;
			
			currentQuestionNum++;
			
			reset();
		}
		
		else {
			this.removeAll();
			this.updateUI();
			
			seinfeldTrivia.getMainFrame().remove(this);
			seinfeldTrivia.getMainFrame().add(scoreboardGui);
			
			seinfeldTrivia.getMainFrame().repaint();
			seinfeldTrivia.getMainFrame().revalidate();
			
			scoreboardGui.loadScoreboard(numCorrect, numIncorrect, numOfQuestions);
		}
			
	}
	
	// Is used when answerTimer runs out and user has not pressed anything
	private void getRightAnswerAndModifyButtons() {
		numIncorrect++;
		
		if(answer1.getText().equals(rightAnswer))
			modifyAnswer1AsCorrect();
		
		else if(answer2.getText().equals(rightAnswer))
			modifyAnswer2AsCorrect();
		
		else if(answer3.getText().equals(rightAnswer))
			modifyAnswer3AsCorrect();
		
		else
			modifyAnswer4AsCorrect();
	}
	
	// resets the whole class, equivalent to inside of constructor
	// Is called on the first QuestionGui/'screen'/question
	private void reset(SeinfeldTrivia seinfeldTrivia, Data data, Graphics graphics, 
			HomeGui homeGui, ScoreboardGui scoreboardGui) {
		this.seinfeldTrivia = seinfeldTrivia;										  
		this.data = data;
		this.graphics = graphics;
		this.homeGui = homeGui;
		this.scoreboardGui = scoreboardGui;
		this.setLayout(new BoxLayout(this, BoxLayout.Y_AXIS));
		
		data.gatherData();
		triviaData = new ArrayList<String>(data.getTriviaData()); // passing Data's triviaData as a parameter 
																  // makes a copy, so the copy in QuestionGui
		                                                          // doesn't affect the original in Data
		questionCount = data.getQuestionCount();                   
		
		questionIndex = new Random().nextInt(questionCount) * 6;
		imgIndex = questionIndex+5;
		
		boundaryIndex1 = questionIndex+1;
		boundaryIndex2 = imgIndex;
		
		correctAnswers = data.getCorrectAnswers();
		rightAnswer = correctAnswers.get(questionIndex/6);
		
		hasAnswered = false;
		
		moveOnDelay = 3; // in seconds
		timeLeft = 10; // default is 10 (in seconds)
		
		currentQuestionNum = 1;
		numOfQuestions = data.getQuestionCount();
		
		answerTimerIsStopped = false;
		
		numCorrect = 0;
		numIncorrect = 0;
		
		initContent();
	}
	
	// resets the whole class, equivalent to inside of constructor, without parameters
	// Is called AFTER program has gone through first QuestionGui/'screen'/question
	private void reset() {
		this.setLayout(new BoxLayout(this, BoxLayout.Y_AXIS));
	
		questionIndex = new Random().nextInt(questionCount) * 6;
		imgIndex = questionIndex+5;
		
		boundaryIndex1 = questionIndex+1;
		boundaryIndex2 = imgIndex;
		
		correctAnswers = data.getCorrectAnswers();
		rightAnswer = correctAnswers.get(questionIndex/6);
		
		hasAnswered = false;

		timeLeft = 10;
		
		answerTimerIsStopped = false;
		
		initContent();
		runAnswerTimer();
	}

	/*
	 * This is part of a thread
	 * 
	 * This is run at the beginning of the program in 
	 * SeinfeldTrivia with the questionThread reference
	 * 
	 * A thread was necessary so the program could continually
	 * check if the user pressed the start trivia button while
	 * simultaneously running the GUI
	 * 
	 * So, the answer timer will not run until the user has
	 * pressed the start trivia button (which is almost
	 * simultaneous with arriving at the QuestionGui 'screen'
	 * or JPanel
	 */
	@Override
	public void run() {
		// while user has not pressed start
		while(!homeGui.getHasPressedStart()) {
			try {
				Thread.sleep(1);
			} catch (InterruptedException e) {
				e.printStackTrace();
			}
			// sleep then try again
		}
		// if has pressed start, initiate the answer timer
		runAnswerTimer();
	}

}


src/elaguy/personal/seinfeldtrivia/main/ScoreboardGui.java
package elaguy.personal.seinfeldtrivia.main;

import java.awt.Component;
import java.awt.Dimension;
import java.awt.Font;
import java.awt.Image;

import javax.swing.Box;
import javax.swing.BoxLayout;
import javax.swing.ImageIcon;
import javax.swing.JLabel;
import javax.swing.JPanel;

public class ScoreboardGui extends JPanel {
	
	private SeinfeldTrivia seinfeldTrivia;
	private Graphics graphics;
	
	private int numCorrect, numIncorrect, numOfQuestions;
	
	private String goldTrophyPath, silverTrophyPath, bronzeTrophyPath;
	
	public ScoreboardGui(SeinfeldTrivia seinfeldTrivia, Graphics graphics) {
		this.seinfeldTrivia = seinfeldTrivia;
		this.graphics = graphics;
		
		this.setLayout(new BoxLayout(this, BoxLayout.Y_AXIS));
		
		goldTrophyPath = "res/gold_trophy.png";
		silverTrophyPath = "res/silver_trophy.png";
		bronzeTrophyPath = "res/bronze_trophy.png";
	}
	
	public void loadScoreboard(int numCorrect, int numIncorrect, int numOfQuestions) {
		this.numCorrect = numCorrect;
		this.numIncorrect = numIncorrect;
		this.numOfQuestions = numOfQuestions;
		
		JLabel youGotLabel = new JLabel("Your Score:");
		youGotLabel.setFont(new Font("Georgia", Font.BOLD, 24));
		youGotLabel.setAlignmentX(Component.CENTER_ALIGNMENT);
		this.add(youGotLabel);
		
		JLabel numCorrectLabel = new JLabel("Questions Correct: " + numCorrect);
		numCorrectLabel.setFont(new Font("Georgia", Font.BOLD, 18));
		numCorrectLabel.setAlignmentX(Component.CENTER_ALIGNMENT);
		this.add(Box.createRigidArea(new Dimension(0, 25)));
		this.add(numCorrectLabel);
		
		JLabel numIncorrectLabel = new JLabel("Questions Incorrect: " + numIncorrect);
		numIncorrectLabel.setFont(new Font("Georgia", Font.BOLD, 18));
		numIncorrectLabel.setAlignmentX(Component.CENTER_ALIGNMENT);
		this.add(Box.createRigidArea(new Dimension(0, 25)));
		this.add(numIncorrectLabel);
		
		JLabel trophyImage = new JLabel();
		ImageIcon imageIcon = chooseTrophyImage();
		Image newImg = graphics.getResizedImage(imageIcon.getImage(), 400, 300);
		imageIcon = new ImageIcon(newImg);
		trophyImage.setIcon(imageIcon);
		trophyImage.setAlignmentX(Component.CENTER_ALIGNMENT);
		this.add(Box.createRigidArea(new Dimension(0, 100)));
		this.add(trophyImage);
		
		seinfeldTrivia.getMainFrame().pack();
	}
	
	private ImageIcon chooseTrophyImage() {
		if(numCorrect == numOfQuestions) // all questions are correct
			return new ImageIcon(goldTrophyPath);
		else if(numCorrect == Math.round((double)numOfQuestions/2)) // around half of the questions are correct
			return new ImageIcon(silverTrophyPath);
		else // less than half or none were correct
			return new ImageIcon(bronzeTrophyPath); 
	}
}


src/elaguy/personal/seinfeldtrivia/main/SeinfeldTrivia.java
package elaguy.personal.seinfeldtrivia.main;

import java.awt.Dimension;

import javax.swing.JButton;
import javax.swing.JFrame;
import javax.swing.UIManager;
import javax.swing.UnsupportedLookAndFeelException;

public class SeinfeldTrivia {
	
	private HomeGui homeGui;
	private QuestionGui questionGui;
	private Data data;
	private Graphics graphics;
	private ScoreboardGui scoreboardGui;
	
	private JFrame mainFrame;
	private final String VERSION = "1.0";

	private final int WIDTH = 800;
	private final int HEIGHT = 600;
	
	public SeinfeldTrivia() {
		
		try {
			UIManager.setLookAndFeel(UIManager.getSystemLookAndFeelClassName());
		} catch (ClassNotFoundException | InstantiationException | IllegalAccessException
				| UnsupportedLookAndFeelException e) {
			e.printStackTrace();
		}
		
		mainFrame = new JFrame("Seinfeld Trivia - Version " + VERSION);
		mainFrame.setMinimumSize(new Dimension(WIDTH, HEIGHT));
		mainFrame.setMaximumSize(new Dimension(WIDTH, HEIGHT));
		mainFrame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
		mainFrame.setVisible(true);
		
		homeGui = new HomeGui(this);
		data = new Data(this);
		graphics = new Graphics(this);
		scoreboardGui = new ScoreboardGui(this, graphics);
		
		questionGui = new QuestionGui(this, data, graphics, homeGui, scoreboardGui);
		Thread questionThread = new Thread(questionGui);
		
		homeGui.setQuestionGui(questionGui);
		data.setQuestionGui(questionGui);
		
		mainFrame.add(homeGui);
		homeGui.setVisible(true);
		questionGui.setVisible(true);
		
		mainFrame.pack();
		
		questionThread.start();
	}				
	
	public JFrame getMainFrame() {
		return mainFrame;
	}
	
	public HomeGui getHomeGui() {
		return homeGui;
	}				

	public int getWIDTH() {
		return WIDTH;
	}

	public int getHEIGHT() {
		return HEIGHT;
	}
	
	public static void main(String[] args) {
		SeinfeldTrivia seinfeldTrivia = new SeinfeldTrivia();
	}

}

src/elaguy/personal/seinfeldtrivia/main/HomeGui.java
package elaguy.personal.seinfeldtrivia.main;

import java.awt.Color;
import java.awt.Component;
import java.awt.Dimension;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import javax.swing.Box;
import javax.swing.BoxLayout;
import javax.swing.ImageIcon;
import javax.swing.JButton;
import javax.swing.JLabel;
import javax.swing.JPanel;

public class HomeGui extends JPanel implements ActionListener {
	
	private SeinfeldTrivia seinfeldTrivia;
	private QuestionGui questionGui;
	
	private boolean hasPressedStart;

	public HomeGui(SeinfeldTrivia seinfeldTrivia) {
		this.seinfeldTrivia = seinfeldTrivia;
		this.setLayout(new BoxLayout(this, BoxLayout.Y_AXIS));
		
		hasPressedStart = false;
		
		initLogo();
		initButtons();
	}
	
	private void initLogo() {
		JLabel seinfeldLogo = new JLabel();
		seinfeldLogo.setIcon(new ImageIcon("res/seinfeld_logo.png"));
		seinfeldLogo.setAlignmentX(Component.CENTER_ALIGNMENT);
		this.add(seinfeldLogo);
		this.add(Box.createRigidArea(new Dimension(0, 125)));
		seinfeldTrivia.getMainFrame().pack();
	}
	
	private void initButtons() {
		JButton startTrivia = new JButton("Start Trivia");
		startTrivia.setForeground(Color.RED);
		startTrivia.setAlignmentX(Component.CENTER_ALIGNMENT);
		startTrivia.setMinimumSize(new Dimension(10, 10));
		startTrivia.setMaximumSize(new Dimension(150, 40));
		startTrivia.addActionListener(this);
		startTrivia.setActionCommand("start");
		startTrivia.setToolTipText("Begin the trivia!");
		this.add(startTrivia);
		
		JButton quit = new JButton("Quit");
		quit.setForeground(Color.RED);
		quit.setAlignmentX(Component.CENTER_ALIGNMENT);
		quit.setMinimumSize(new Dimension(10, 10));
		quit.setMaximumSize(new Dimension(150, 40));
		this.add(Box.createRigidArea(new Dimension(0, 45)));
		quit.addActionListener(this);
		quit.setActionCommand("quit");
		quit.setToolTipText("Quit the program.");
		this.add(quit);
		
		seinfeldTrivia.getMainFrame().pack();
	}
	
	public void actionPerformed(ActionEvent e) {
		if(e.getActionCommand().equals("start")) {
			this.removeAll();
			this.updateUI();
			
			seinfeldTrivia.getMainFrame().remove(this);
			seinfeldTrivia.getMainFrame().add(questionGui);
		
			seinfeldTrivia.getMainFrame().repaint();
			seinfeldTrivia.getMainFrame().revalidate();
			
			hasPressedStart = true;
		}
		
		else if(e.getActionCommand().equals("options")) {
			this.removeAll();
			this.updateUI();
			
			seinfeldTrivia.getMainFrame().remove(this);
			
			seinfeldTrivia.getMainFrame().repaint();
			seinfeldTrivia.getMainFrame().revalidate();
		}
		
		else {
			seinfeldTrivia.getMainFrame().dispose();
			System.exit(0);
		}
	}
	
	public void setQuestionGui(QuestionGui questionGui) {
		this.questionGui = questionGui;
	}
	
	public boolean getHasPressedStart() {
		return hasPressedStart;
	}
}